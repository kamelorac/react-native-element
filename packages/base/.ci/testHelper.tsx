import React, { JSXElementConstructor } from 'react';
import { View } from 'react-native';
import {
  render,
  RenderOptions,
  fireEvent,
  act,
} from '@testing-library/react-native';

export { fireEvent, act };

// for getting findByType e.g. wrapper.findByType(Icon) see implementation in Avatar Component
export const renderWithWrapper = (
  children: React.ReactElement<any, string | JSXElementConstructor<any>>,
  wrapperTestID?: string,
  _themeProp: unknown = {},
  renderOptions?: RenderOptions
) => {
  const options: RenderOptions = {
    ...(!wrapperTestID && {
      wrapper: (props) => <View {...props} testID="wrapper" />,
    }),
    ...renderOptions,
  };
  const renderApi = render(children, options);
  const wrapper = renderApi.queryByTestId(wrapperTestID || 'wrapper')!;
  return { wrapper, ...renderApi };
};
