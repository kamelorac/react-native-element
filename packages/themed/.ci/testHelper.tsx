import React, { JSXElementConstructor } from 'react';
import { View } from 'react-native';
import {
  render,
  RenderOptions,
  fireEvent,
  act,
} from '@testing-library/react-native';
import {
  ThemeProvider,
  FullTheme,
  createTheme,
  CreateThemeOptions,
} from '../src/config';

export { fireEvent, act };

// for getting findByType e.g. wrapper.findByType(Icon) see implementation in Avatar Component
export const renderWithWrapper = (
  children: React.ReactElement<any, string | JSXElementConstructor<any>>,
  wrapperTestID?: string,
  themeProp: Partial<CreateThemeOptions> = {},
  renderOptions?: RenderOptions
) => {
  const options: RenderOptions = {
    ...(!wrapperTestID && {
      wrapper: (props) => <View {...props} testID="wrapper" />,
    }),
    ...renderOptions,
  };
  const renderApi = render(
    <ThemeProvider theme={createTheme(themeProp)}>{children}</ThemeProvider>,
    options
  );
  const wrapper = renderApi.queryByTestId(wrapperTestID || 'wrapper')!;
  return { wrapper, ...renderApi };
};
