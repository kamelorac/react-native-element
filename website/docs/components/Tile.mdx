---
id: tile
title: Tile
---

import { IconsStyle } from "@site/src/components/Docs_Icons";
import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";
import CodeBlock from "@theme/CodeBlock";
import { BiInfoCircle } from "react-icons/bi";
import Usage from "../component_usage/Tile.mdx";
import Play from "@site/playground/Tile/tile.playground";

<IconsStyle />

Tiles like Cards, are a convenient way to display related content about a single subject.
Also receives all [TouchableNativeFeedback](http://reactnative.dev/docs/touchablenativefeedback.html#props) (Android) or [TouchableOpacity](http://reactnative.dev/docs/touchableopacity.html#props) (iOS) props.

## Usage

<div class="row inline-flex-center">
  <div class="col col--3">
    <h4>Import</h4>
  </div>
  <div class="col col--9">
    <CodeBlock>{`import { Tile } from '@rneui/themed';`}</CodeBlock>
  </div>
  <div class="col col--3">
    <h4>
      Theme Key{" "}
      <a href="../customizing#using-themeprovider">
        <BiInfoCircle />
      </a>
    </h4>
  </div>
  <div class="col col--9">
    <CodeBlock>{`Tile`}</CodeBlock>
  </div>
</div>

<Usage />

## Props

:::note
Includes all [View](https://reactnative.dev/docs/view#props) props.
:::

<div class='table-responsive'>

| Name                    | Type                              | Default          | Description                                                                     |
| ----------------------- | --------------------------------- | ---------------- | ------------------------------------------------------------------------------- |
| `ImageComponent`        | typeof Component                  | `RNE Image`      | Custom ImageComponent for Tile.                                                 |
| `activeOpacity`         | number                            |                  | Number passed to control opacity on press.                                      |
| `caption`               | ReactNode                         |                  | Text inside the tilt when tile is featured.                                     |
| `captionStyle`          | Text Style                        |                  | Styling for the caption (optional); You only use this if `caption` is a string. |
| `containerStyle`        | View Style                        |                  | Styling for the outer tile container.                                           |
| `contentContainerStyle` | View Style                        |                  | Styling for bottom container when not featured tile.                            |
| `featured`              | boolean                           |                  | Changes the look of the tile.                                                   |
| `height`                | number                            | `'width/2'`      | Height for the tile.                                                            |
| `icon`                  | IconObject                        |                  | Icon Component Props.                                                           |
| `iconContainerStyle`    | View Style                        |                  | Styling for the outer icon container.                                           |
| `imageContainerStyle`   | View Style                        |                  | Styling for the image.                                                          |
| `imageProps`            | ImageProps(Object)                | `{}`             | Optional properties to pass to the image if provided e.g "resizeMode".          |
| `imageSrc`              | `string` \| `ImageSourcePropType` |                  | Source for the image.                                                           |
| `overlayContainerStyle` | View Style                        |                  | Styling for the overlay container when using featured tile.                     |
| `title`                 | string                            |                  | Text inside the tile.                                                           |
| `titleNumberOfLines`    | number                            |                  | Number of lines for Title.                                                      |
| `titleStyle`            | Text Style                        |                  | Styling for the title.                                                          |
| `width`                 | number                            | `'screen width'` | Width for the tile.                                                             |

</div>

## Playground

<Play />
