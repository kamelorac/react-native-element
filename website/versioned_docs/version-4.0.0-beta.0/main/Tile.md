---
id: tile

title: Tile

slug: /tile
---

import Usage from './usage/Tile/Tile.md'

Tiles like Cards, are a convenient way to display related content about a single subject.

Also receives all [TouchableNativeFeedback](http://reactnative.dev/docs/touchablenativefeedback.html#props) (Android) or [TouchableOpacity](http://reactnative.dev/docs/touchableopacity.html#props) (iOS) props.

## Usage

<Usage />

---

## Props

### Tile

- [ImageComponent](#imagecomponent)
- [activeOpacity](#activeopacity)
- [caption](#caption)
- [captionStyle](#captionstyle)
- [containerStyle](#containerstyle)
- [contentContainerStyle](#contentcontainerstyle)
- [featured](#featured)
- [height](#height)
- [icon](#icon)
- [iconContainerStyle](#iconcontainerstyle)
- [imageContainerStyle](#imagecontainerstyle)
- [imageProps](#imageprops)
- [imageSrc](#imagesrc)
- [overlayContainerStyle](#overlaycontainerstyle)
- [title](#title)
- [titleNumberOfLines](#titlenumberoflines)
- [titleStyle](#titlestyle)
- [width](#width)

## Reference

### Tile

#### ImageComponent

Custom ImageComponent for Tile.

| Type | Default |
| ---- | ------- |
| any  | None    |

---

#### activeOpacity

Number passed to control opacity on press.

| Type   | Default |
| ------ | ------- |
| number | None    |

---

#### caption

Text inside the tilt when tile is featured.

| Type | Default |
| ---- | ------- |
| any  | None    |

---

#### captionStyle

Styling for the caption (optional); You only use this if `caption` is a string.

| Type               | Default |
| ------------------ | ------- |
| Text Style(Object) | None    |

---

#### containerStyle

Styling for the outer tile container.

| Type               | Default |
| ------------------ | ------- |
| View style(Object) | None    |

---

#### contentContainerStyle

Styling for bottom container when not featured tile.

| Type               | Default |
| ------------------ | ------- |
| View style(Object) | None    |

---

#### featured

Changes the look of the tile.

| Type    | Default |
| ------- | ------- |
| boolean | None    |

---

#### height

Height for the tile.

| Type   | Default      |
| ------ | ------------ |
| number | width \* 0.8 |

---

#### icon

Icon Component Props.

| Type               | Default |
| ------------------ | ------- |
| View style(Object) | None    |

---

#### iconContainerStyle

Styling for the outer icon container.

| Type               | Default |
| ------------------ | ------- |
| View style(Object) | None    |

---

#### imageContainerStyle

Styling for the image.

| Type               | Default |
| ------------------ | ------- |
| View style(Object) | None    |

---

#### imageProps

Optional properties to pass to the image if provided e.g "resizeMode".

| Type               | Default |
| ------------------ | ------- |
| ImageProps(Object) | {}      |

---

#### imageSrc

Source for the image.

| Type                          | Default |
| ----------------------------- | ------- |
| string or ImageSourcePropType | None    |

---

#### overlayContainerStyle

Styling for the overlay container when using featured tile.

| Type               | Default |
| ------------------ | ------- |
| View style(Object) | None    |

---

#### title

Text inside the tile.

| Type   | Default |
| ------ | ------- |
| string | None    |

---

#### titleNumberOfLines

Number of lines for Title.

| Type   | Default |
| ------ | ------- |
| number | None    |

---

#### titleStyle

Styling for the title.

| Type               | Default |
| ------------------ | ------- |
| Text Style(Object) | None    |

---

#### width

Width for the tile.

| Type   | Default                        |
| ------ | ------------------------------ |
| number | Dimensions.get('window').width |

---
