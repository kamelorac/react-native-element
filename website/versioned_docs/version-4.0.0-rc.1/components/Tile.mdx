---
id: tile
title: Tile
---

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";
import {} from "react-native-elements";
import Usage from "../component_usage/Tile.mdx";

Tiles like Cards, are a convenient way to display related content about a single subject.
Also receives all [TouchableNativeFeedback](http://reactnative.dev/docs/touchablenativefeedback.html#props) (Android) or [TouchableOpacity](http://reactnative.dev/docs/touchableopacity.html#props) (iOS) props.

## Usage

<Usage />

## Props

:::note
Includes all [View](https://reactnative.dev/docs/view#props) props.
:::

<div class='table-responsive'>

| Name                    | Type                              | Default          | Description                                                                     |
| ----------------------- | --------------------------------- | ---------------- | ------------------------------------------------------------------------------- |
| `ImageComponent`        | typeof Component                  | `RNE Image`      | Custom ImageComponent for Tile.                                                 |
| `activeOpacity`         | number                            |                  | Number passed to control opacity on press.                                      |
| `caption`               | ReactNode                         |                  | Text inside the tilt when tile is featured.                                     |
| `captionStyle`          | Text Style                        |                  | Styling for the caption (optional); You only use this if `caption` is a string. |
| `containerStyle`        | View Style                        |                  | Styling for the outer tile container.                                           |
| `contentContainerStyle` | View Style                        |                  | Styling for bottom container when not featured tile.                            |
| `featured`              | boolean                           |                  | Changes the look of the tile.                                                   |
| `height`                | number                            | `'width/2'`      | Height for the tile.                                                            |
| `icon`                  | IconObject                        |                  | Icon Component Props.                                                           |
| `iconContainerStyle`    | View Style                        |                  | Styling for the outer icon container.                                           |
| `imageContainerStyle`   | View Style                        |                  | Styling for the image.                                                          |
| `imageProps`            | ImageProps(Object)                | `{}`             | Optional properties to pass to the image if provided e.g "resizeMode".          |
| `imageSrc`              | `string` or `ImageSourcePropType` |                  | Source for the image.                                                           |
| `overlayContainerStyle` | View Style                        |                  | Styling for the overlay container when using featured tile.                     |
| `title`                 | string                            |                  | Text inside the tile.                                                           |
| `titleNumberOfLines`    | number                            |                  | Number of lines for Title.                                                      |
| `titleStyle`            | Text Style                        |                  | Styling for the title.                                                          |
| `width`                 | number                            | `'screen width'` | Width for the tile.                                                             |

</div>
