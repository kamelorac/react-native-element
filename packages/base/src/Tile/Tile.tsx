import React from 'react';
import {
  View,
  StyleSheet,
  Dimensions,
  Pressable,
  PressableProps,
  ViewStyle,
  StyleProp,
  TextStyle,
  ImageProps,
  ImageURISource,
  ImageSourcePropType,
} from 'react-native';
import { Image } from '../Image';
import { Text } from '../Text';
import { Icon, IconObject, IconProps } from '../Icon';
import { FeaturedTile } from './components/FeaturedTile';
import { androidRipple, defaultTheme, RneFunctionComponent } from '../helpers';
import Color from 'color';

export interface TileProps extends PressableProps {
  /** Text inside the tile. */
  title?: string;

  /** Icon Component Props.
   * @type IconObject
   */
  icon?: IconObject & IconProps;

  /** Text inside the tilt when tile is featured. */
  caption?: React.ReactNode;

  /** Source for the image. */
  imageSrc?: ImageSourcePropType | string | number;

  /** Number passed to control opacity on press. */
  activeOpacity?: number;

  /** Styling for the outer tile container. */
  containerStyle?: StyleProp<ViewStyle>;

  /** Styling for the image. */
  imageContainerStyle?: StyleProp<ViewStyle>;

  /** Styling for the outer icon container. */
  iconContainerStyle?: StyleProp<ViewStyle>;

  /** Styling for the overlay container when using featured tile. */
  overlayContainerStyle?: StyleProp<ViewStyle>;

  /** Styling for the title. */
  titleStyle?: StyleProp<TextStyle>;

  /** Styling for the caption (optional); You only use this if `caption` is a string. */
  captionStyle?: StyleProp<TextStyle>;

  /** Width for the tile.
   * @default 'screen width'
   */
  width?: number;

  /** Height for the tile.
   * @default 'width/2'
   */
  height?: number;

  /** Changes the look of the tile. */
  featured?: boolean;

  /** Styling for bottom container when not featured tile. */
  contentContainerStyle?: StyleProp<ViewStyle>;

  /** Number of lines for Title. */
  titleNumberOfLines?: number;

  /** Optional properties to pass to the image if provided e.g "resizeMode". */
  imageProps?: Partial<ImageProps>;

  /** Custom ImageComponent for Tile.
   * @default RNE Image
   */
  ImageComponent?: typeof React.Component;
}

/** Tiles like Cards, are a convenient way to display related content about a single subject.
 * Also receives all [TouchableNativeFeedback](http://reactnative.dev/docs/touchablenativefeedback.html#props) (Android) or [TouchableOpacity](http://reactnative.dev/docs/touchableopacity.html#props) (iOS) props. */
export const Tile: RneFunctionComponent<TileProps> = ({
  featured,
  imageSrc,
  icon,
  title,
  children,
  caption,
  titleStyle,
  onPress,
  activeOpacity,
  overlayContainerStyle,
  captionStyle,
  iconContainerStyle,
  imageContainerStyle,
  containerStyle,
  contentContainerStyle,
  titleNumberOfLines,
  ImageComponent = Image,
  imageProps = {},
  width = Dimensions.get('window').width,
  height = width * 0.8,
  theme = defaultTheme,
  ...attributes
}) => {
  if (featured) {
    const featuredProps = {
      title,
      icon,
      caption,
      imageSrc,
      onPress,
      activeOpacity,
      containerStyle,
      imageContainerStyle,
      overlayContainerStyle,
      titleStyle,
      captionStyle,
      width,
      height,
      imageProps,
      ImageComponent,
    };
    return <FeaturedTile {...(featuredProps as TileProps)} />;
  }

  return (
    <Pressable
      {...{
        onPress,
        android_ripple: androidRipple(
          Color(theme?.colors?.primary).alpha(activeOpacity).rgb().toString()
        ),
        ...attributes,
      }}
      style={StyleSheet.flatten([
        {
          width,
          height,
        },
        containerStyle && containerStyle,
      ])}
    >
      <ImageComponent
        resizeMode="cover"
        source={imageSrc as ImageURISource}
        containerStyle={StyleSheet.flatten([
          styles.imageContainer,
          imageContainerStyle && imageContainerStyle,
        ])}
        style={{
          ...StyleSheet.absoluteFillObject,
          alignItems: 'center',
          justifyContent: 'center',
        }}
        {...(imageProps as Partial<ImageProps>)}
      >
        <View
          style={StyleSheet.flatten([
            styles.iconContainer,
            iconContainerStyle && iconContainerStyle,
          ])}
        >
          {icon && <Icon {...icon} />}
        </View>
      </ImageComponent>

      <View
        style={StyleSheet.flatten([
          styles.contentContainer,
          contentContainerStyle && contentContainerStyle,
        ])}
      >
        <Text
          testID="tileTitle"
          h4={!titleStyle || !('fontSize' in titleStyle)}
          style={StyleSheet.flatten([styles.text, titleStyle && titleStyle])}
          numberOfLines={titleNumberOfLines}
        >
          {title}
        </Text>
        {children}
      </View>
    </Pressable>
  );
};

const styles = StyleSheet.create({
  imageContainer: {
    flex: 2,
  },
  text: {
    backgroundColor: 'rgba(0,0,0,0)',
    marginBottom: 5,
  },
  contentContainer: {
    paddingTop: 15,
    paddingBottom: 5,
    paddingLeft: 15,
    paddingRight: 15,
  },
  iconContainer: {
    justifyContent: 'center',
    alignItems: 'center',
    alignSelf: 'center',
  },
});

Tile.displayName = 'Tile';
