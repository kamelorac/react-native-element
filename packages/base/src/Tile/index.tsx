import { Tile, TileProps } from './Tile';

export { Tile };
export type { TileProps };
