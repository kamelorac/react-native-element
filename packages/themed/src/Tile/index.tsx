import { withTheme } from '../config';
import { Tile, TileProps } from '@rneui/base/dist/Tile/Tile';
import { FeaturedTile as BaseFeaturedTile } from '@rneui/base/dist/Tile/components/FeaturedTile';

export type { TileProps };
export const FeaturedTile = withTheme<TileProps>(
  BaseFeaturedTile,
  'FeaturedTile'
);
export default withTheme<TileProps>(Tile, 'Tile');
