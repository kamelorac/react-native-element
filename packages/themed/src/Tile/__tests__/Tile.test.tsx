import React from 'react';
import Tile from '../index';
import { renderWithWrapper } from '../../../.ci/testHelper';
import { Text } from 'react-native';

describe('Tile component', () => {
  it('should render', () => {
    const { toJSON } = renderWithWrapper(
      <Tile imageSrc={{ uri: 'http://google.com' }} />
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('should apply styles from theme', () => {
    const theme = {
      Tile: {
        title: 'Mary is friendly',
      },
    };
    const { wrapper } = renderWithWrapper(
      <Tile imageSrc={{ uri: 'http://google.com' }} activeOpacity={0.9} />,
      '',
      { components: theme }
    );
    expect(wrapper.findByType(Text).props.children).toBe('Mary is friendly');
  });
});
