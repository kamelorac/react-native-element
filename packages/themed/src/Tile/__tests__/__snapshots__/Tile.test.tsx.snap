// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tile component should render 1`] = `
<View
  testID="wrapper"
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    replaceTheme={[Function]}
    style={
      {
        "height": 600,
        "width": 750,
      }
    }
    updateTheme={[Function]}
  >
    <View
      accessibilityIgnoresInvertColors={true}
      style={
        {
          "backgroundColor": "transparent",
          "flex": 2,
          "overflow": "hidden",
          "position": "relative",
        }
      }
    >
      <Image
        onLoad={[Function]}
        resizeMode="cover"
        source={
          {
            "uri": "http://google.com",
          }
        }
        style={
          {
            "alignItems": "center",
            "bottom": 0,
            "justifyContent": "center",
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
        testID="RNE__Image"
        transitionDuration={360}
      />
      <View
        accessibilityElementsHidden={true}
        collapsable={false}
        importantForAccessibility="no-hide-descendants"
        pointerEvents="none"
        style={
          {
            "bottom": 0,
            "left": 0,
            "opacity": 1,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#bdbdbd",
              "bottom": 0,
              "justifyContent": "center",
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
          testID="RNE__Image__placeholder"
        />
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "bottom": 0,
            "justifyContent": "center",
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
        testID="RNE__Image__children__container"
      >
        <View
          style={
            {
              "alignItems": "center",
              "alignSelf": "center",
              "justifyContent": "center",
            }
          }
        />
      </View>
    </View>
    <View
      style={
        {
          "paddingBottom": 5,
          "paddingLeft": 15,
          "paddingRight": 15,
          "paddingTop": 15,
        }
      }
    >
      <Text
        accessibilityRole="text"
        style={
          {
            "backgroundColor": "rgba(0,0,0,0)",
            "color": "#242424",
            "fontSize": 28.285714285714285,
            "marginBottom": 5,
          }
        }
        testID="tileTitle"
      />
    </View>
  </View>
</View>
`;
